### PR TITLE
Let supervisor run with :logger option

### PR DESCRIPTION
**Pull request by nownabe**
Hi, thank you for a nice tool! I like it :heart: 
I found one problem when using [Sneakers](https://github.com/jondot/sneakers) with supervisor.

When Supervisor is given a Logger instance as :logger option, an exception occurs ([detail](https://github.com/nownabe/serverengine-playground/blob/master/debug_20150521/51b60fa_output_3.txt)):

```
Unexpected error comparison of Fixnum with String failed
```
And when Supervisor is given both :logger option and :log option, another exception occurs ([detail](https://github.com/nownabe/serverengine-playground/blob/master/debug_20150521/51b60fa_output_4.rb)):
```
Unexpected error undefined method `logdev=' for #<Logger:0x007fc745189d18>
```

I used [this](https://github.com/nownabe/serverengine-playground/blob/master/debug_20150521/test.rb) check script.

These errors occur because ServerEngine::DaemonLogger has richer methods than Logger and Supervisor use these methods.

In this pull request, if given logger is not a ServerEngine::DaemonLogger, ConfigLoader ignore :logger option.
This is my proposal.

Please check it.

### CODE REVIEW

**Comment by swatanabeidcf** [comment]
Can latest v1.5.10 pass CI with Ruby 2.1.2?
In my local, test failed with 2.1.2.
With 2.1.6, test passed.

**Comment by nownabe** [comment]
swatanabeidcf is my another account for work...

**Comment by repeatedly** [comment]
Thanks for the PR. I will check it later.

**Comment by repeatedly** [comment]
Could you remove .travis.yml related commit?
I just updated .travis.yml for testing lastest versions and using proper command.

**Comment by nownabe** [comment]
Thank you for checking CI.
I removed the commit.

Should I cherry-pick your patch?

**Comment by repeatedly** [comment]
> Should I cherry-pick your patch?

It is better :)

**Comment by nownabe** [comment]
Ok. Test passed, thanks!

File: lib/serverengine/config_loader.rb [hunk 1]
```diff
@@ -56,8 +56,8 @@ def reload_config
     private
 
     def create_logger
-      if logger = @config[:logger]
-        @logger = logger
+      if @config[:logger].is_a?(DaemonLogger)
+        @logger = @config[:logger]
         return
```
Comment (frsyuki) [inline lib/serverengine/config_loader.rb]: `create_logger` needs to return a logger. I think this is the bug. So, I think create_logger method needs to be:

```ruby
    def create_logger
      if logger = @config[:logger]
        @logger = logger
      else
        @logger = @logger_class.new(logdev_from_config(@config), @config)
      end
      return @logger
    end
```

Comment (frsyuki) [inline lib/serverengine/config_loader.rb]: `undefined method `logdev=' for` is another issue.

Solution would be:

```

**Comment by frsyuki** [comment]
`:logger` was not expected. `:logger_class` was expected. Please keep in mind that `:logger` option may not work with `worker_type=spawn`.

**Comment by nownabe** [comment]
Thank you for your feedback! :smile: 
I understood. I will feed back to Sneakers and fix my PR.

1) I have a question before that, logger class given with `:logger_class` has to have `reopen!` method?
Is it unneccessary to check that logger class have required methods?
I concern that server will crash because of logger.

2) `DaemonLogger#level=` can receive String because this method covert String to Integer Constat like `Logger::INFO`.
But general `Logger#level=` cannot convert.
In this situation, `Unexpected error comparison of Fixnum with String failed` occurs.

I think it is more safe to use only `DaemonLogger` or child class of `DaemonLogger`.
How would you like to restrict logger class?

**Comment by frsyuki** [comment]
@nownabe Thanks!

> 1) logger class given with :logger_class has to have reopen! method?

Yes. But it is OK to do nothing on the method.

> 2) DaemonLogger#level= can receive String because this method covert String to Integer Constat like Logger::INFO. But general Logger#level= cannot convert.

That's right. That is a good point that ServerEngine should somehow improve it.

One idea is that ServerEngine (especially `ConfigLoader#create_logger`) wraps an instantiated logger class with a proxy class that implements `reopen!` and `level=` methods.

Another idea is to propose `Logger.reopen!` method to Ruby community. I'll try send idea.

**Comment by nownabe** [comment]
@frsyuki Thank you for your review! :bow::bow::bow:
That proxy class looks so good!
And I think Ruby's `Logger#reopen!` method is useful. I want it too. :yum:

I will fix my code in this PR.

**Comment by frsyuki** [comment]
`::Logger#level=:symbol` is proposed to Ruby by @sonots:
* https://github.com/ruby/ruby/pull/1101
* https://bugs.ruby-lang.org/issues/11695

**Comment by tagomoris** [comment]
Let me close this. This pull-request has not been updated for long time after last review comment.

**Pull request closed without merging**